langchain4j-neo4j/src/main/java/dev/langchain4j/store/embedding/neo4j/Neo4jEmbeddingStore.java

```diff
@@ -0,0 +1,244 @@
+package dev.langchain4j.store.embedding.neo4j;
+
+import dev.langchain4j.data.embedding.Embedding;
+import dev.langchain4j.data.segment.TextSegment;
+import dev.langchain4j.store.embedding.EmbeddingMatch;
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import lombok.Builder;
+import lombok.Getter;
+import org.neo4j.driver.AuthTokens;
+import org.neo4j.driver.GraphDatabase;
+import org.neo4j.driver.Session;
+import org.neo4j.driver.Values;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.stream.IntStream;
+
+import static dev.langchain4j.internal.Utils.getOrDefault;
+import static dev.langchain4j.internal.Utils.isCollectionEmpty;
+import static dev.langchain4j.internal.Utils.randomUUID;
+import static dev.langchain4j.internal.ValidationUtils.*;
+import static java.util.Collections.singletonList;
+
+import org.neo4j.driver.Driver;
+import org.neo4j.driver.SessionConfig;
+
+import static dev.langchain4j.store.embedding.neo4j.Neo4jEmbeddingUtils.*;
+
+/**
+ * Represents a Vector index as an embedding store.
+ * Annotated with `@Getter` to be used in {@link Neo4jEmbeddingUtils}
+ */
+@Getter
+public class Neo4jEmbeddingStore implements EmbeddingStore<TextSegment> {
+
+    private static final Logger log = LoggerFactory.getLogger(Neo4jEmbeddingStore.class);
+
+    /* Neo4j Java Driver settings */
+    private final Driver driver;
+    private final SessionConfig config;
+
+    /* Neo4j schema field settings */
+    private final int dimension;
+    private final Neo4jDistanceType distanceType;
+
+    private final String indexName;
+    private final String metadataPrefix;
+    private final String embeddingProperty;
+    private final String label;
+    private final String text;
+
+    /**
+     * Creates an instance of Neo4jEmbeddingStore defining a {@link Driver} 
+     * starting from uri, user and password
+     */
+    public static class Neo4jEmbeddingStoreBuilder {
+        public Neo4jEmbeddingStoreBuilder withBasicAuth(String uri, String user, String password) {
+            return this.driver(GraphDatabase.driver(uri, AuthTokens.basic(user, password)));
+        }
+    }
+
+    /**
+     * Creates an instance of Neo4jEmbeddingStore
+     * @param driver: the {@link Driver} (required)
+     * @param dimension: the dimension (required)
+     * @param config: the {@link SessionConfig}  (optional, default is `SessionConfig.forDatabase("neo4j")`)
+     * @param label: the optional label name (default: "Document")
+     * @param property: the optional property name (default: "embedding")
+     * @param distanceType: the optional distanceType (default: "cosine")
+     * @param metadataPrefix: the optional metadata prefix (default: "")
+     * @param text: the optional text property name (default: "text")
+     * @param indexName: the optional index name (default: "langchain-embedding-index")
+     */
+    @Builder
+    public Neo4jEmbeddingStore(
+            SessionConfig config,
```

Comment:
Good question if we want to leak the session config from neo4j into the client code. I would probably start with database name and then add a builder methods for other options.

Should the builder constructor be public?

Reply:
Added databaseName config as well